src/gui/iconutils.cpp

```diff
@@ -1,52 +1,140 @@
+#include <QLoggingCategory>
 #include <QPainter>
 #include <QPixmapCache>
 #include <QSvgRenderer>
 
+namespace {
+QString findSvgFilePath(const QString &fileName, const QStringList &possibleColors)
+{
+    const QString baseSvgNoColor{QString{OCC::Theme::themePrefix} + fileName};
+    if (QFile::exists(baseSvgNoColor)) {
+        return baseSvgNoColor;
+    }
+
+    for (const auto &color : possibleColors) {
+        const QString baseSVG{QString{OCC::Theme::themePrefix} + color + QLatin1Char('/') + fileName};
+
+        if (QFile::exists(baseSVG)) {
+            return baseSVG;
+        }
+    }
+
+    return {};
+}
+}
+
 namespace OCC {
 namespace Ui {
-namespace IconUtils {
-QPixmap pixmapForBackground(const QString &fileName, const QColor &backgroundColor)
-{
-    Q_ASSERT(!fileName.isEmpty());
+    Q_LOGGING_CATEGORY(lcIconUtils, "nextcloud.gui.iconutils", QtInfoMsg)
+    namespace IconUtils {
+        QPixmap pixmapForBackground(const QString &fileName, const QColor &backgroundColor)
+        {
+            Q_ASSERT(!fileName.isEmpty());
+
+            const auto pixmapColor = backgroundColor.isValid()
+                && !Theme::isDarkColor(backgroundColor)
+                ? QColorConstants::Svg::black
+                : QColorConstants::Svg::white;
+
+            return createSvgPixmapWithCustomColor(fileName, pixmapColor);
+        }
+
+        QPixmap createSvgPixmapWithCustomColor(const QString &fileName, const QColor &customColor, const QSize &size)
+        {
+            Q_ASSERT(!fileName.isEmpty());
+            Q_ASSERT(customColor.isValid());
+
+            if (fileName.isEmpty()) {
+                qWarning(lcIconUtils) << "fileName is empty";
+            }
+
+            if (!customColor.isValid()) {
+                qWarning(lcIconUtils) << "customColor is invalid";
+            }
+
+            const auto customColorName = customColor.name();
+
+            const QString cacheKey = fileName + QLatin1Char(',') + customColorName;
 
-    // some icons are present in white or black only, so, we need to check both when needed
-    const auto iconBaseColors = QStringList({ QStringLiteral("black"), QStringLiteral("white") });
+            QPixmap cachedPixmap;
 
-    const QString pixmapColor = backgroundColor.isValid() && !Theme::isDarkColor(backgroundColor) ? "black" : "white";
+            // check for existing QPixmap in cache
+            if (QPixmapCache::find(cacheKey, &cachedPixmap)) {
+                return cachedPixmap;
+            }
+
+            // some icons are present in white or black only, so, we need to check both when needed
+            const auto iconBaseColors = QStringList{QStringLiteral("black"), QStringLiteral("white")};
+
+            // check if there is an existing pixmap matching the custom color
+            if (iconBaseColors.contains(customColorName)) {
+                cachedPixmap = QPixmap::fromImage(QImage{QString{OCC::Theme::themePrefix} + customColorName + QLatin1Char('/') + fileName});
```

Comment:
@FlexW What is the benefit?

Reply:
@FlexW I've checked the rest of the codebase. We use `QLatin1Char` everywhere where we need just one character created at compile time. Even some of the most recent code is using it. What would be the benefit of using `QStringLiteral` instead?

Reply:
`QLatin1Char` is weird as we are supposed to be using `UTF-8` encoding for our source files and I am not sure there is an optimized override that would speed up the string concatenation by using it
the whole point of `QLatin1...` classes is to use them only when an optimized override exists in the API otherwise they are slower than normal `QString` (or for a constant `QStringLiteral`)